base/delete_soon_helper.h

```diff
@@ -0,0 +1,52 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_BASE_DELETE_SOON_HELPER_H_
+#define BRAVE_BASE_DELETE_SOON_HELPER_H_
+
+namespace base {
+// Helper that forces object to invalidate things before object destroyed
+// Subclass must implement DeleteSoonImpl and call DeleteSoon before destructor
+// of the object. Example usage would be like a object is a WeakPtr which is
+// called and destructed on different threads, in this case we can
+// DeleteSoonHelper to make sure the WeakPtr is invalidate on certain thread.
+// See tor::TorFileWatcher as a reference.
+template <typename T>
+class DeleteSoonHelper {
```

Comment:
@bridiver Frankly speaking, I think the helper makes things more complicated than they are

Reply:
I don't understand how this makes things more complicated. https://github.com/brave/brave-core/pull/7927#discussion_r584102841 is incorrect, it does provide guarantees and that is the reason it exists. The entire reason for use base::Bind in the first place (vs std::bind) is to prevent common errors and this prevents another class of common errors in the code.

Reply:
addressed in https://github.com/brave/brave-core/pull/7927/commits/536325b0aca2613cdb3ca3e1c5deac2a25f7dfd9